Add a wrapper class for datasets to work with sklearn

BenjaminBossan opened the pull request:
When wrapping a torch dataset with `SliceDatasetX`, it can be passed to sklearn grid search et al.

Fixes #443

taketwo commented on docs/user/helper.rst:
```diff
@@ -19,6 +19,35 @@ length of the arrays and not the number of keys, and you get a
 with :class:`.SliceDict`, this works.
 
 
+.. _slicedictx:
```
The class name is `SliceDatasetX`, but the anchor name has _dict_ in it. Is this on purpose?

BenjaminBossan replied:
Thanks, this was unintended

taketwo commented on docs/user/helper.rst:
```diff
@@ -19,6 +19,35 @@ length of the arrays and not the number of keys, and you get a
+-------------
+
+A :class:`.SliceDatasetX` is a wrapper for
+:class:`torch.utils.data.Dataset`\s that makes them behave a little
```
I think in order to avoid convoluted names in the running text, our convention was to use unqualified name and spell PyTorch in front, i.e. in this case:

> is a wrapper for PyTorch :class:`~torch.utils.data.Dataset`s

BenjaminBossan replied:
Will fix it.

thomasjpfan commented on skorch/helper.py:
```diff
@@ -131,6 +132,82 @@ def __ne__(self, other):
+    def __getitem__(self, i):
+        if isinstance(i, (int, np.integer)):
+            Xi = _unpack_first(self.dataset[i])
+            return self.transform(Xi)
```
Currently, `self.transform` returns `X`, do we intend on letting users to subclass `SliceDatasetX` to provide their own transform?

BenjaminBossan replied:
That's the intent. I will document it better.

BenjaminBossan commented:
I made several improvements to the implementation.

* I updated docstrings and docs according to the reviewer comments above.
* I changed the implementation to make it easy to also wrap y in this class (and hence renamed it from `SliceDatasetX` to `SliceDataset`.
* I noticed that with the old implementation, when data is e.g. split into train and valid durng grid search, all data is loaded into memory, which we don't want. Now the implementation is closer to a PyTorch `Subset`. I.e. when a `SliceDataset` is indexed by a list, array, or slice, it creates a `SliceDataset` subset that doesn't actually load any data. Only when it is indexed by an int will it load the data. This avoids the mentioned issue.
Of course, this makes the assumption that when sklearn touches the data, it will never slice using an int. AFAIK, this assumption should hold, but sklearn makes no guarantees about that. If anyone knows a more robust solution, please chime in. What we need is basically a way to know during slicing whether the actual data is needed at this moment or not.

BenjaminBossan commented:
@ottonemo Could you review this pls?

thomasjpfan commented on skorch/helper.py:
```diff
@@ -131,6 +132,105 @@ def __ne__(self, other):
+    dataset : torch.utils.data.Dataset
+      A valid torch dataset.
+
+    n : int (default=0)
```
Is there a better variable name for this?

BenjaminBossan replied:
I have no good idea, any suggestions?

githubnemo replied:
How about renaming the parameter to `key` to be in line with builtins like `sorted` and `max`?

BenjaminBossan replied:
The argument doesn't serve the same purpose, though, since `key` in those functions is a callable.

A more similar function would be `itemgetter`, where the argument is called `item`. But `itemgetter` also has `items` for returning a tuple, so now I wonder if we should just use `itemgetter` under the hood to make it possible to return multiple items.

thomasjpfan commented on skorch/helper.py:
```diff
@@ -131,6 +132,105 @@ def __ne__(self, other):
+        self.indices = indices
+
+        self.indices_ = (self.indices if self.indices is not None
+                         else np.arange(len(self.dataset)))
```
Is the `indices=None` case tested?

BenjaminBossan replied:
Yes, as it's the default. But I added a test for `indices != None`

thomasjpfan commented on skorch/tests/test_helper.py:
```diff
@@ -254,6 +254,189 @@ def test_equals_different_keys(self, sldict_cls):
+        [-3],
+        [0, 10, 3, -8, 3],
+        np.ones(100, dtype=np.bool),
+        np.array([0, 0, 1, 0] * 25, dtype=np.bool),
```
`* 25` and cast to bool?

thomasjpfan replied:
Oops. Noticed `[0, 0, 1, 0]` is a list, (Not an numpy array)

thomasjpfan commented on skorch/helper.py:
```diff
@@ -131,6 +132,105 @@ def __ne__(self, other):
+
+    def __getitem__(self, i):
+        if isinstance(i, (int, np.integer)):
+            Xi = self.dataset[self.indices_[i]][self.n]
```
Do we want a custom error message when `self.n` out of bounds?

thomasjpfan commented on skorch/tests/test_helper.py:
```diff
@@ -254,6 +254,189 @@ def test_equals_different_keys(self, sldict_cls):
+        (np.array([0, 0, 1, 0] * 25, dtype=np.bool), 6),
+    ])
+    def test_slice_twice(self, slds_cls, custom_ds, X, y, sl0, sl1, n):
+        data = y if n else X
```
`n==1` is slightly clearer, but this is okay.

thomasjpfan commented on skorch/tests/test_helper.py:
```diff
@@ -254,6 +254,189 @@ def test_equals_different_keys(self, sldict_cls):
 
 
+class TestSliceDataset:
+    @pytest.fixture
```
```suggestion
    @pytest.fixture(scope='class')
```

BenjaminBossan commented:
@thomasjpfan I addressed your comments

BenjaminBossan commented:
@ottonemo Is this good to go?

githubnemo commented on skorch/tests/test_helper.py:
```diff
@@ -254,6 +254,209 @@ def test_equals_different_keys(self, sldict_cls):
+        class MyDataset(Dataset):
+            """Simple pytorch dataset that returns 2 values"""
+            def __init__(self, X, y):
+                super().__init__(X, y)
```
Why have the `__init__` implementation at all?

BenjaminBossan replied:
True, I will remove this.

githubnemo commented:
@BenjaminBossan I'm currently out of office until Friday so I can't merge it but there wasn't anything except for the awkward name of parameter `n` for which I just made an alternative suggestion (but we can keep it if there are objections to my suggestion) and my question regarding `__init__`.

BenjaminBossan commented:
So would `item` be better than `n`?

thomasjpfan commented:
`index` works too. (I am okay with `item` or `index`)

BenjaminBossan commented:
My first association with `index` would be the `index` method on strings, lists, tuples, but maybe that's just me. I could see `idx` though, that might be better than `item`.

thomasjpfan commented:
It is indexing a tuple returned by the `Dataset`. Let's do `idx`?
